File: adv14.js
function RecipeCalculator(initialRecipe) {
    this.board = initialRecipe;
    this.cursor1 = 0;
    this.cursor2 = 1;
}

RecipeCalculator.prototype.nextRecipe = function() {
        let first = parseInt(this.board[this.cursor1]);
        let second = parseInt(this.board[this.cursor2]);

        let sum = first + second;
        if (sum < 10) {
            this.board += sum;
        } else {
            this.board += (Math.floor(sum / 10)).toString() + (sum % 10).toString();
        }

        this.cursor1 = (this.cursor1 + first + 1) % this.board.length;
        this.cursor2 = (this.cursor2 + second + 1) % this.board.length;
}

RecipeCalculator.prototype.calculateRecipes = function (iterations) {
    while (this.board.length < iterations) {
        this.nextRecipe();
    }
}

RecipeCalculator.prototype.calculateRecipe = function (recipes) {
    let iterations = recipes + 10 - this.board.length;

    this.calculateRecipes(iterations);
    return this.board.substr(recipes, 10);
}

const step = Math.pow(2, 15);
RecipeCalculator.prototype.findRecipe = function(recipe) {
    let position;
    while ((position = this.board.indexOf(recipe)) == -1) {
        lastPosition = this.board.length - 1;
        for (let i = 0; i < step; i++)
            this.nextRecipe();
    }

    return position;
}

let recipe = new RecipeCalculator("37");

console.log("Part One:", recipe.calculateRecipe(260321));


console.log(recipe.findRecipe("260321"));

File: adv14.cpp
#include <iostream>
#include <cstdlib>
#include <cstring>

using namespace std;

class RecipeCalculator {
    public:
        RecipeCalculator(string initialBoard) {
            board = initialBoard;
            board.reserve(1048576);
        }
        void nextRecipe() {
            int first = board[firstCursor] - '0';
            int second = board[secondCursor] - '0';

            int sum = first + second;

            
            if (sum >= 10) {
                board.append(1, char(sum / 10 + '0'));
            }

            board.append(1, sum % 10 + '0');

            firstCursor = (firstCursor + first + 1) % board.length();
            secondCursor = (secondCursor + second + 1) % board.length();
        }

        const string calculateRecipe (int recipes) {
            int targetLength = recipes + 10;

            while (board.length() < targetLength)
                nextRecipe();

            return board.substr(recipes, 10);
        }

        const int step = 1024 * 1024;
        int findRecipe(const string &recipe) {
            int prevStep = 0;
            int found = -1;
            while ((found = board.find(recipe)) == -1) {
                calculateRecipe(board.length() + step);

                if (board.length() - prevStep > 1000000) {
                    prevStep = board.length();
                }
            }
            return found;
        }

        const string &getBoard() {
            return board;
        }

    private:
        string board;
        int firstCursor = 0, secondCursor = 1;
};

int main() {
    RecipeCalculator recipe("37");

    cout << "Part One: " << recipe.calculateRecipe(260321) << endl;
    cout << "Part Two: " << recipe.findRecipe("260321") << endl;
    return 0;
}
